csr/docker/launch.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python3
+
+import datetime
+import logging
+import os
+import random
+import re
+import signal
+import subprocess
+import sys
+import telnetlib
+import time
+
+import vrnetlab
+
+def handle_SIGCHLD(signal, frame):
+    os.waitpid(-1, os.WNOHANG)
+
+def handle_SIGTERM(signal, frame):
+    sys.exit(0)
+
+signal.signal(signal.SIGINT, handle_SIGTERM)
+signal.signal(signal.SIGTERM, handle_SIGTERM)
+signal.signal(signal.SIGCHLD, handle_SIGCHLD)
+
+TRACE_LEVEL_NUM = 9
+logging.addLevelName(TRACE_LEVEL_NUM, "TRACE")
+def trace(self, message, *args, **kws):
+    # Yes, logger takes its '*args' as 'args'.
+    if self.isEnabledFor(TRACE_LEVEL_NUM):
+        self._log(TRACE_LEVEL_NUM, message, args, **kws)
+logging.Logger.trace = trace
+
+
+
+class CSR_vm(vrnetlab.VM):
+    def __init__(self, username, password, install_mode=False):
+        for e in os.listdir("/"):
+            if re.search(".qcow2$", e):
+                disk_image = "/" + e
+        super(CSR_vm, self).__init__(username, password, disk_image=disk_image)
+
+        self.install_mode = install_mode
+        self.num_nics = 10
+
+        if self.install_mode:
+            logger.trace("install mode")
+            self.image_name = "config.iso"
+            self.create_boot_image()
+            time.sleep(1)
+
+            self.qemu_args.extend(["-cdrom", "/" +self.image_name]) 
+
+
+    def create_boot_image(self):
+        """ Creates a iso image with a bootstrap configuration
+        """
+
+        self.cfg_file = open('/iosxe_config.txt', 'w')
+        self.cfg_file.write("platform console serial\r\n\r\n")
```

Comment:
FYI, you can set the newline to `\r\n` in the open() call - see https://docs.python.org/3/library/io.html#io.open and specifically the `newline` argument. This way you can pretend it's just a newline and say `\n` in your write() but it will actually put `\r\n` in the output file.